Increase testing coverage of rcutils to 95%

brawner opened the pull request:
I've added a bit more test cases to the unit tests in rcutils to boost the coverage to 95% (just barely).

The big question about this PR is the desired location of `time_bomb_allocator`. @hidmic had suggested putting it into a centralized location so it can be used in other packages. I'm thinking I might want to do that for the allocator_testing_utils.h as well. But I don't think rcutils/include is the desired end location.

The remaining 5% of untested logic pretty much consists of failure checks that won't fail, generally because bad parameters have already been sanitized by the time the function under test calls a function that is checked for failure. In a few cases it might be because the system call cannot be made to fail reliably or even at all in a simple unit test.

* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=10965)](http://ci.ros2.org/job/ci_linux/10965/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=6305)](http://ci.ros2.org/job/ci_linux-aarch64/6305/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=8927)](http://ci.ros2.org/job/ci_osx/8927/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=10875)](http://ci.ros2.org/job/ci_windows/10875/)
Signed-off-by: Stephen Brawner <brawner@gmail.com>

hidmic commented on test/test_get_env.cpp:
```diff
@@ -49,4 +50,16 @@ TEST(TestGetEnv, test_get_env) {
+
+  const char * home = NULL;
+  const char * ret = rcutils_get_env("HOME", &home);
+  EXPECT_EQ(NULL, ret);
```
@brawner mind to explain why can it be assumed that `HOME` is not in the environment?

brawner replied:
The return of `rcutils_get_env` is an error string. If it's null then the call succeeded. The actual variable value is assigned to `home`.

hidmic replied:
Oops, nevermind.

cottsay replied:
What is the point of this test though? Won't it fail if `HOME` isn't set on Windows, for example?

brawner replied:
The original intent of this check was to verify that "HOME" was defined, so that we can be sure that the expected post condition below was a result of setting each environment variable to NULL. However, I forgot to update this when I added in the #ifdef for windows.

I have since updated this to check the appropriate HOME/USERPROFILE variables in the ifdef below.

hidmic commented on test/test_logging_bad_env.cpp:
```diff
@@ -0,0 +1,22 @@
+
+TEST(TestBadCustomEnv, test_initialize) {
+  EXPECT_FALSE(g_rcutils_logging_initialized);
+  ASSERT_EQ(RCUTILS_RET_INVALID_ARGUMENT, rcutils_logging_initialize());
```
@brawner uh, `RCUTILS_RET_INVALID_ARGUMENT` for a function that takes no arguments?

brawner replied:
They are set in the CMakeLists.txt file. But `rcutils_logging_initialize()` checks environment variables and will return an invalid argument in that case.

hidmic replied:
I see. It's out of scope for this PR, but I find the error code selection weird. Environment variables are not function arguments.

hidmic commented on test/test_logging_console_output_handler.cpp:
```diff
@@ -0,0 +1,94 @@
+    EXPECT_EQ(RCUTILS_RET_OK, rcutils_logging_shutdown());
+  });
+
+  rcutils_log_location_t log_location = {
```
@brawner what follows seems akin to a smoke test to me. Shall we leave a comment stating so? Same for the other test case below.

brawner replied:
Added a more informative comment at the top of the test function.

hidmic replied:
Cool. Mocks would come in handy here...

hidmic commented on test/test_repl_str.cpp:
```diff
@@ -62,4 +62,31 @@ TEST(test_repl_str, nominal) {
+    for (size_t i = 0; i < 1048576; ++i) {
+      ss << "f";
+    }
+    char * out = rcutils_repl_str(ss.str().c_str(), "f", "longer replacementment", &allocator);
```
@brawner typo? 
```suggestion
    char * out = rcutils_repl_str(ss.str().c_str(), "f", "longer replacement", &allocator);
```

brawner replied:
Fixed.

hidmic commented on test/test_strerror.cpp:
```diff
@@ -43,4 +43,12 @@ TEST(test_strerror, get_error) {
+  // Hopefully this does not become a valid errno.
+  errno = 12345;
+  rcutils_strerror(error_string, sizeof(error_string));
+  ASSERT_STREQ(error_string, "Failed to get error");
```
@brawner a message to hint our future selves in case of failure would be nice.

brawner replied:
Added a message.

brawner commented:
Separate PR to fix Windows bug:
https://github.com/ros2/rcutils/pull/259

brawner commented:
After merging #259, I've rebased this PR. Also for API considerations I moved the time_bomb allocator into the test directory. We can leave that discussion for post-foxy discussions.

brawner commented:
Retesting:
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=11009)](http://ci.ros2.org/job/ci_linux/11009/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=6344)](http://ci.ros2.org/job/ci_linux-aarch64/6344/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=8968)](http://ci.ros2.org/job/ci_osx/8968/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=10935)](http://ci.ros2.org/job/ci_windows/10935/)

brawner commented:
Also without #257 this shows up as only 94%. At 1666/1763, it just needs one more line of coverage or one less line of uncovered code 🤷‍♂️

cottsay commented on test/test_filesystem.cpp:
```diff
@@ -241,6 +244,16 @@ TEST_F(TestFilesystemFixture, is_readable_and_writable) {
+    {
+      g_allocator.deallocate(path, g_allocator.state);
+    });
+    ASSERT_FALSE(nullptr == path);
```
```suggestion
    ASSERT_NE(nullptr, path);
```

brawner replied:
After private discussion, leaving this as is to match the rest the file.

brawner commented:
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=11011)](http://ci.ros2.org/job/ci_linux/11011/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=6346)](http://ci.ros2.org/job/ci_linux-aarch64/6346/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=8970)](http://ci.ros2.org/job/ci_osx/8970/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=10938)](http://ci.ros2.org/job/ci_windows/10938/)

hidmic commented on test/test_logging_console_output_handler.cpp:
```diff
@@ -0,0 +1,97 @@
+
+// There are no outputs of the handler function, and the only result are fprintf() calls.
+// This is just a smoke test to check that the code can handle simple inputs cleanly.
+TEST(TestLoggingConsoleOutputHandler, typical_inputs) {
```
@brawner meta: thinking again about this test, you _could_ mock `stdout` and/or `stderr` for the duration of the test. Won't block on it, but it is doable.

brawner replied:
I tried this out and couldn't get it to work. Assigning a new file pointer to stdout/stderr left me with blank files. Alternatively, using `freopen()` worked but it wouldn't let me reset stdout/stderr back in an easily portable way.

I like the idea of checking expected input on stdout/stderr, but I'll need a better solution, which may have to come later.

hidmic replied:
Hmm, interesting. Oh well, it can be done in a follow-up PR.

hidmic commented on test/test_logging_console_output_handler.cpp:
```diff
@@ -0,0 +1,97 @@
+
+// There are no outputs of the handler function, and the only result are fprintf() calls.
+// This is just a smoke test to check that the code can handle bad inputs cleanly.
+TEST(TestLoggingConsoleOutputHandler, bad_inputs) {
```
@brawner same as above.

brawner commented:
Testing once more before merging:
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=11049)](http://ci.ros2.org/job/ci_linux/11049/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=6373)](http://ci.ros2.org/job/ci_linux-aarch64/6373/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=9000)](http://ci.ros2.org/job/ci_osx/9000/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=10972)](http://ci.ros2.org/job/ci_windows/10972/)
